### DISCUSSION TITLE
Was VM GB Resource Output Rolled Back?

### DISCUSSION THREAD

**Discussion by szollin** [netbox-community/netbox#11586] [Q&A] [answered]
I am trying to get the VM memory resource to display as GB. On PR 6700, issue 6667 was merged for 2.11.8 which was a while back. This allowed such function. On issue 7083, a visual error with the merge was fixed so it displayed the units correctly. As of right now (3.4.1), this looks like it isn't possible. I am trying to find out if this change was rolled back at one point and, if so, why was it rolled back?

**Comment by candlerb** [comment, marked as answer]
Can you be more specific about your problem?  Please show an example of what you see, and what you expected to see instead.

The code is all still there in v3.4.3:

```
./netbox/utilities/templatetags/helpers.py:def humanize_megabytes(mb):
./netbox/templates/virtualization/virtualmachine.html:                                {{ object.memory|humanize_megabytes }}
./netbox/templates/virtualization/cluster.html:                          {{ memory_sum|humanize_megabytes }}
```

The logic is that you always enter the VM RAM size in MiB, but:
* If you store a size which is a multiple of 1024, it is divided by 1024 and displayed as "GB" (_sic_ - officially should be "GiB")
* If you store a size which is a multiple of 1024^2, it is divided by 1024^2 and displayed as "TB" (ditto)

```
def humanize_megabytes(mb):
    """
    Express a number of megabytes in the most suitable unit (e.g. gigabytes or terabytes).
    """
    if not mb:
        return ''
    if not mb % 1048576:  # 1024^2
        return f'{int(mb / 1048576)} TB'
    if not mb % 1024:
        return f'{int(mb / 1024)} GB'
    return f'{mb} MB'
```

Note that a value like 1536 will still display as 1536 MB, not as 1.5 GB, by design. (If you want to query that design decision, it will have to go into a feature request with a proposal of *exactly* how you think it should work instead)

Given the logic above, it works fine for me: `/virtualization/virtual-machines/NNN/`

![image](https://user-images.githubusercontent.com/44789/214798069-3d8c9844-b25c-4c4e-87de-5bd5bb65f124.png)

although in the columnar list of VMs it still displays as MB, which the column heading makes clear: `/virtualization/virtual-machines/`

![image](https://user-images.githubusercontent.com/44789/214797858-1059e66d-d25d-4846-9fa3-ef451519c761.png)

**Reply by szollin** [reply]
Thanks for the full explanation @candlerb. Inside the VM is fine and I can see the GB numbers without issue there. The issue is with the columnar view. MB is actually on the low end of common VM memory sizes. I was under the impression that the PRs made changes to the columnar view. Is it possible to choose column header measurements?

**Reply by candlerb** [reply]
> I was under the impression that the PRs made changes to the columnar view.

Can you point to where you got that impression?

Issue #6667 was resolved by commit 18a9e39b which only updates the VM details view page. And #7083 was resolved by 65c93396 which only removed the spurious "MB" string.

**Reply by szollin** [reply]
I got the impression from the proposed functionality request on #6667.

> Change the displayed unit of VM memory from MB (megabyte) to GB (gigabyte) or make the unit selectable when creating / editing the VM.


Since I can not change the values in the create screen, this lead to an understanding of the display being GB. To me, display is anywhere displayed.

**Reply by candlerb** [reply]
I see problems with changing the column display to work in the same way: in particular, display values like "2 GB" and "1536 MB" are unlikely to sort correctly.

**Reply by szollin** [reply]
I get where your at on that. Multiple programs have that issue. Even "1.5 GB" will sort weird.

Thanks for your help on understanding this more. I will mark your explanation as an answer.
